src/aleph/http.clj

```diff
@@ -358,96 +362,106 @@
+                                               (-> (conn' req)
+                                                   (maybe-timeout! request-timeout)
+
+                                                   ;; request timeout triggered
+                                                   (d/catch' TimeoutException
+                                                             (fn [^Throwable e]
+                                                               (d/error-deferred (RequestTimeoutException. e))))
+
+                                                   ;; clean up the connection
+                                                   (d/chain'
+                                                    (fn cleanup-conn [rsp]
+
+                                                      ;; either destroy/dispose of the conn, or release it back for reuse
+                                                      (-> (:aleph/destroy-conn? rsp)
+                                                          (maybe-timeout! read-timeout)
+
+                                                          ;; read timeout triggered
+                                                          (d/catch' TimeoutException
+                                                                    (fn [^Throwable e]
+                                                                      (log/trace "Request timed out.")
+                                                                      (d/error-deferred (ReadTimeoutException. e))))
+
+                                                          (d/chain'
+                                                           (fn [early?]
+                                                             (if (or early?
+                                                                     (not (:aleph/keep-alive? rsp))
+                                                                     (<= 400 (:status rsp)))
+                                                               (do
+                                                                 (log/trace "Connection finished. Disposing...")
+                                                                 (flow/dispose pool k conn))
+                                                               (flow/release pool k conn)))))
+                                                      (-> rsp
+                                                          (dissoc :aleph/destroy-conn?)
+                                                          (assoc :connection-time (- end start)))))))))
+
+                                         (fn handle-response [rsp]
+                                           (->> rsp
+                                                (middleware/handle-cookies req)
+                                                (middleware/handle-redirects request req))))))))))))
+                      req))]
+      (d/connect response result)
```

Comment:
I'm not sure to understand, why do we need to `connect` here?
We cannot just `catch'` on the `response` deferred?

Reply:
Yeah this is a bit subtle: In the context of this patch it's for [handling this race](https://github.com/clj-commons/aleph/blob/75c4cbf3904026d1c4223f8588850c52be9115dd/src/aleph/http.clj#L389-L392) since the `response` deferred cannot refer to itself. In the follow-up patch for cancellation of connection establishment, we will also need to connect the result deferred to `(first conn)` to support it, which again would require `response` to refer to itself... Hope that makes sense!

Reply:
Got it, thanks for the explanation.